cpp/src/request_tag.cpp

```diff
@@ -113,25 +115,26 @@ void RequestTag::request()
 
   if (_delayedSubmission->_transferDirection == TransferDirection::Send) {
     param.cb.send = tagSendCallback;
-    request       = ucp_tag_send_nbx(_endpoint->getHandle(),
-                               _delayedSubmission->_buffer,
-                               _delayedSubmission->_length,
-                               _delayedSubmission->_data.getTag()._tag,
-                               &param);
+    auto tagSend  = _delayedSubmission->_data.getTagSend();
+    request       = ucp_tag_send_nbx(
+      _endpoint->getHandle(), tagSend._buffer, tagSend._length, tagSend._tag, &param);
```

Comment:
Because the `_data` is now a tagged union, it has the information needed to do the dispatch.

There are a few ways to do this:

```
auto data = delayedSubmission->data;
if (std::holds_alternative<TagSend>(data)) {
    auto info = data.getTagSend();
   ...
else if (std::holds_alternative<TagRecv>(data)) {
   auto info = data.getTagRecv();
   ...
} else {
   throw ucxx::logic_error("Invalid delayed submission for tag send/recv"); // or whatever the message...
}
```

Note the problem here is that we must unwrap things manually. However, we can instead use `std::visit` with operator overloading:

```
struct Dispatch {
   void operator()(TagSend const &ts) { ucp_tag_send_nbx(...); };
   void operator()(TagRecv const &tr) { ucp_tag_recv_nbx(...); };
   template<typename T>
   void operator()(T const &) { throw std::logic_error("Unreachable!"): }
}
...

std::visit(Dispatch{}, data);
```

Or, my favourite since you don't need to build a struct, but needs some template magic https://www.cppstories.com/2019/02/2lines3featuresoverload.html/

```c++
template <class... Ts> struct overload : Ts... {
  using Ts::operator()...;
};
template <class... Ts> overload(Ts...) -> overload<Ts...>;

void RequestTag::request()
{
  ucp_request_param_t param = ...;
  void * request = std::visit(overload {
    [&param](TagSend const &ts) -> void * { param.cb.send = tagSendCallback; return ucp_tag_send_nbx(...); },
    [&param](TagRecv const &tr) -> void * { param.cb.recv = tagRecvCallback; return ucp_tag_recv_nbx(...); },
    [](auto const &) -> void * { throw std::logic_error("Unreachable!"); }
    }, delayedSubmission->data);
}
```

Basically, this is poor-man's pattern matching with pretty ugly syntax. But, it becomes a compile error if you don't handle all variant types, good! It also keeps everything in one place.